java/core/src/main/java/com/microsoft/bond/StructBondType.java

```diff
@@ -0,0 +1,1359 @@
+        this.serializeStructFields(context, value);
+        context.writer.writeBaseEnd();
+    }
+
+    @Override
+    protected final TStruct deserializeValue(TaggedDeserializationContext context) throws IOException {
+        TStruct value = this.newDefaultValue();
+        context.reader.readStructBegin();
+        if (this.baseStructType != null) {
+            this.baseStructType.deserializeValueHelperForBase(context, value);
+        }
+        this.deserializeStructFields(context, value);
+        context.reader.readStructEnd();
+        return value;
+    }
+
+    // recursive helper
+    private void deserializeValueHelperForBase(
+            TaggedDeserializationContext context, TStruct value) throws IOException {
+        if (this.baseStructType != null) {
+            this.baseStructType.deserializeValueHelperForBase(context, value);
+        }
+        context.reader.readBaseBegin();
+        this.deserializeStructFields(context, value);
+        context.reader.readBaseEnd();
+    }
+
+    @Override
+    protected final void serializeField(
+            SerializationContext context,
+            TStruct value,
+            StructField<TStruct> field) throws IOException {
+        this.verifySerializedNonNullableFieldIsNotSetToNull(value, field);
+        // struct fields are never omitted
+        context.writer.writeFieldBegin(BondDataType.BT_STRUCT, field.getId(), field);
+        try {
+            this.serializeValue(context, value);
+        } catch (InvalidBondDataException e) {
+            // throws
+            Throw.raiseStructFieldSerializationError(false, field, e, null);
+        }
+        context.writer.writeFieldEnd();
+    }
+
+    @Override
+    protected final TStruct deserializeField(
+            TaggedDeserializationContext context,
+            StructField<TStruct> field) throws IOException {
+        // a struct value may be deserialized only from BT_STRUCT
+        if (context.readFieldResult.type.value != BondDataType.BT_STRUCT.value) {
+            // throws
+            Throw.raiseFieldTypeIsNotCompatibleDeserializationError(context.readFieldResult.type, field);
+        }
+        TStruct value = null;
+        try {
+            value = this.deserializeValue(context);
+        } catch (InvalidBondDataException e) {
+            // throws
+            Throw.raiseStructFieldSerializationError(true, field, e, null);
+        }
+        return value;
+    }
+
+    @Override
+    final boolean equalsInternal(BondType<?> obj) {
+        // the caller makes sure that the class of the argument is the same as the class of this object
+        StructBondType that = (StructBondType) obj;
+        if (this.genericTypeSpecialization != null) {
+            return this.genericTypeSpecialization.equals(that.genericTypeSpecialization);
+        } else {
+            return that.genericTypeSpecialization == null;
+        }
+    }
+
+    // private top-level serialization entry point
```

Comment:
It is package private, which is effectively private to Bond core. It's not visible in public API.

---
In reply to: [123399250](https://github.com/Microsoft/bond/pull/502#discussion_r123399250) [](ancestors = 123399250)